libcst/codemod/commands/rename.py

```diff
@@ -0,0 +1,170 @@
+# Copyright (c) Facebook, Inc. and its affiliates.
+#
+# This source code is licensed under the MIT license found in the
+# LICENSE file in the root directory of this source tree.
+#
+# pyre-strict
+import argparse
+from typing import Optional, Sequence, Union
+
+import libcst as cst
+from libcst.codemod import CodemodContext, VisitorBasedCodemodCommand
+from libcst.codemod.visitors import AddImportsVisitor, RemoveImportsVisitor
+from libcst.helpers import get_full_name_for_node
+from libcst.metadata import QualifiedName, QualifiedNameProvider, QualifiedNameSource
+
+
+class RenameCommand(VisitorBasedCodemodCommand):
+    """
+    Rename all instances of a local or imported object.
+    """
+
+    DESCRIPTION: str = "Rename all instances of a local or imported object."
+
+    METADATA_DEPENDENCIES = (QualifiedNameProvider,)
+
+    @staticmethod
+    def add_args(parser: argparse.ArgumentParser) -> None:
+        parser.add_argument(
+            "--orig_module",
+            dest="orig_module",
+            help="Name of the module to rename. Leave blank if renaming a local variable.",
+        )
+        parser.add_argument(
+            "--orig_object",
+            dest="orig_object",
+            required=True,
+            help="Name of the object in module or variable to rename.",
+        )
+        parser.add_argument(
+            "--new_module",
+            dest="new_module",
+            help="New name of the module. Leave blank if renaming a local variable, or if same as orig_module.",
+        )
+        parser.add_argument(
+            "--new_object",
+            dest="new_object",
+            required=True,
+            help="New name of the object in module or variable.",
+        )
```

Comment:
I wonder about simplifying this so that we just take `oldname` and `newname`, where either one or both can be a dotted path. If either one contains one or more dots, we internally split it on the final dot to split module name from object name. Otherwise the logic remains the same.

So instead of typing `--orig_module foo.bar --orig_object baz --new_module one.two --new_object three` I would just type `--oldname foo.bar.baz --newname one.two.three`, which is a lot less verbose and I think clearer to read also.

One question then might be whether the meaning of `--oldname foo.bar.baz --newname bing` should be shorthand for `--oldname foo.bar.baz --newname foo.bar.bing` (that's what the current logic would suggest) or should instead be "rename baz to bing and remove import of baz from foo.bar without replacing it" implying that it's becoming module-local. I would favor the latter, I think it's kinda non-obvious for omitting new-module to mean "don't change old module".

Reply:
This could be a good solution to disambiguate the codemod ie provide the full qualified name of the object one is renaming. One scenario I can see arising though is what about the case where we just want to rename a module eg `foo.bar.baz` where `baz` is still a module and not an object. But perhaps we could just do two checks in the codemod for each node we encounter and see if one of them is a match...

Reply:
It's clearer if we talk about which syntactic forms we want to support rather than what `baz` _is_, because the codemod doesn't actually know or care whether `baz` in `foo.bar.baz` is a module or some other kind of object; even in the current form if you say `--orig_module foo.bar --orig_object baz`, `baz` could in fact be a module but the codemod will still happily change `from foo.bar import baz` to the requested new name. That's part of the reason I think it's more intuitive and will lead to fewer surprises if we just ask for a single dotted name and replace any syntactic form we see that matches it, rather than making implicit assertions about what is or isn't a module.

So I'm guessing the syntactic form you mean here is if we have `import foo.bar.baz` or `from foo.bar.baz import qux, quux` and we want to change `foo.bar.baz` to something else, without changing any names imported from it or accessed as attributes on it? Yes, I think in my API proposal if we specified `--oldname foo.bar.baz` that case should be supported.